File: src/lsp/schema-body.go
// The body of a note against its schema: which chapters a schema wants, the
// order it names them in, and the sections standing under a numbered one. The
// frontmatter's own checker stands in schema.go beside this file.
// [[spec/design_output/schema#a-finding-names-the-section]]
package main

import (
	"quackitect/yaml"

	"fmt"
	"strconv"
	"strings"
)

type standingAt struct {
	Section
	at int
}

func bodyFaults(note Note, spec *yaml.Doc, kind, where string) []Finding {
	level := 1
	if spec.Has("headingLevel") {
		level = yaml.AsInt(spec.Get("headingLevel"))
	}
	wanted := chaptersWanted(yaml.AsList(spec.Get("sections")), note.Front.Said)

	standing := []standingAt{}
	for i, one := range note.Sections {
		if one.Level == level {
			standing = append(standing, standingAt{Section: one, at: i})
		}
	}

	named := map[string]*yaml.Doc{}
	for _, one := range wanted {
		if rule := yaml.AsDoc(one); rule != nil {
			named[yaml.AsString(rule.Get("header"))] = rule
		}
	}

	out := []Finding{}
	for _, one := range wanted {
		rule := yaml.AsDoc(one)
		if rule == nil || !yaml.AsBool(rule.Get("required")) {
			continue
		}
		header := yaml.AsString(rule.Get("header"))
		if headed(standing, header) >= 0 {
			continue
		}
		out = append(out, schemaFault(header, where, 1,
			fmt.Sprintf("A %s note carries a %s chapter.", kind, header)))
	}

	for _, held := range standing {
		if _, ours := named[held.Header]; ours {
			continue
		}
		if spec.Get("extraSections") == false {
			out = append(out, schemaFault(held.Header, where, held.Line,
				fmt.Sprintf("The %s schema names no %s chapter.", kind, held.Header)))
		}
	}

	if yaml.AsString(spec.Get("order")) == "strict" {
		out = append(out, orderFaults(standing, wanted, kind, where)...)
	}
	out = append(out, lastFaults(standing, wanted, where)...)

	for _, held := range standing {
		if rule, ours := named[held.Header]; ours {
			out = append(out, sectionFaults(held, rule, note, where)...)
		}
	}
	return out
}

// [[spec/design_output/schema#three-keywords-name-a-step]]
func chaptersWanted(sections []any, front *yaml.Doc) []any {
	out := []any{}
	for _, one := range sections {
		rule := yaml.AsDoc(one)
		list := ""
		if rule != nil {
			list = yaml.AsString(rule.Get("x-one-per"))
		}
		if list == "" {
			out = append(out, one)
			continue
		}
		if front == nil {
			continue
		}
		for _, step := range yaml.AsList(front.Get(list)) {
			said := yaml.AsDoc(step)
			if said == nil {
				continue
			}
			header := strings.TrimSpace(yaml.AsString(said.Get("name")))
			if header == "" {
				continue
			}
			chapter := yaml.New()
			chapter.Set("header", header)
			chapter.Set("required", true)
			out = append(out, chapter)
		}
	}
	return out
}

func orderFaults(standing []standingAt, wanted []any, kind, where string) []Finding {
	order := []string{}
	for _, one := range wanted {
		if rule := yaml.AsDoc(one); rule != nil {
			order = append(order, yaml.AsString(rule.Get("header")))
		}
	}

	held := []standingAt{}
	for _, one := range standing {
		if at(order, one.Header) >= 0 {
			held = append(held, one)
		}
	}

	out := []Finding{}
	for i := 1; i < len(held); i++ {
		if at(order, held[i].Header) > at(order, held[i-1].Header) {
			continue
		}
		out = append(out, schemaFault(held[i].Header, where, held[i].Line,
			fmt.Sprintf("%s stands after %s, and a %s note puts it first.", held[i].Header, held[i-1].Header, kind)))
	}
	return out
}

func lastFaults(standing []standingAt, wanted []any, where string) []Finding {
	out := []Finding{}
	for _, one := range wanted {
		rule := yaml.AsDoc(one)
		if rule == nil || yaml.AsString(rule.Get("position")) != "last" {
			continue
		}
		header := yaml.AsString(rule.Get("header"))
		found := headed(standing, header)
		if found < 0 || found == len(standing)-1 {
			continue
		}
		out = append(out, schemaFault(header, where, standing[found].Line,
			fmt.Sprintf("%s closes this note, and %s stands after it.", header, standing[found+1].Header)))
	}
	return out
}

func sectionFaults(held standingAt, rule *yaml.Doc, note Note, where string) []Finding {
	out := []Finding{}
	items := itemsIn(held.Own)

	if yaml.AsBool(rule.Get("list")) && len(items) == 0 {
		out = append(out, schemaFault(held.Header, where, held.Line,
			fmt.Sprintf("%s holds a list of items.", held.Header)))
	}
	if yaml.AsBool(rule.Get("ordered")) && someUnnumbered(items) {
		out = append(out, schemaFault(held.Header, where, held.Line,
			fmt.Sprintf("%s numbers every item.", held.Header)))
	}
	if most := yaml.AsInt(rule.Get("maxItems")); most > 0 && len(items) > most {
		out = append(out, schemaFault(held.Header, where, held.Line+items[most].line,
			fmt.Sprintf("A note holds %d items.", most)))
	}
	if spec := yaml.AsDoc(rule.Get("subsections")); spec != nil {
		out = append(out, underFaults(held, spec, note, where)...)
	}
	return out
}

type numberedSection struct {
	said   int
	line   int
	header string
}

func underFaults(held standingAt, spec *yaml.Doc, note Note, where string) []Finding {
	level := held.Level + 1
	if spec.Has("headingLevel") {
		level = yaml.AsInt(spec.Get("headingLevel"))
	}

	out := []Finding{}
	numbers := []numberedSection{}
	for _, one := range note.Sections[held.at+1:] {
		if one.Level <= held.Level {
			break
		}
		if one.Level != level {
			continue
		}
		found := numberedAt.FindStringSubmatch(one.Header)
		if yaml.AsBool(spec.Get("numbered")) && found == nil {
			out = append(out, schemaFault(held.Header, where, one.Line,
				fmt.Sprintf("A chapter under %s opens with the number of the item it argues.", held.Header)))
			continue
		}
		if found != nil {
			said, _ := strconv.Atoi(found[1])
			numbers = append(numbers, numberedSection{said: said, line: one.Line, header: one.Header})
		}
	}

	if yaml.AsString(spec.Get("order")) != "strict" {
		return out
	}
	for i := 1; i < len(numbers); i++ {
		if numbers[i].said > numbers[i-1].said {
			continue
		}
		out = append(out, schemaFault(held.Header, where, numbers[i].line,
			fmt.Sprintf("%s stands after %s, and the numbers run up.", numbers[i].header, numbers[i-1].header)))
	}
	return out
}

File: .claude/skills/level0/lib/schema-body.js
// The body checks: the chapters a schema wants, their order, the list and
// the table under a chapter, and the placeholder a mint leaves.
// [[spec/design_output/schema#a-schema-names-its-chapters]]

import { fault, LEFT } from "./schema-fault.js";
import { FENCE, readNote } from "./schema-read.js";
import { tableFaults } from "./schema-table.js";

const ITEM = /^\s*(?:\d+[.)]|[-*+])\s+\S/;

const NUMBERED = /^(\d+)[.)]?\s/;

const COMMENT = /<!--[\s\S]*?-->/g;

export function bodyFaults(note, spec, kind, where) {
  const level = spec.headingLevel ?? 1;
  const wanted = chaptersWanted(spec.sections ?? [], note.front.said ?? {}, level);
  const levels = new Set(wanted.map((one) => one.level ?? level));
  const nested = wanted.some((one) => (one.level ?? level) > level);
  const standing = note.sections.filter((one) =>
    nested ? one.level >= level : levels.has(one.level),
  );
  const top = standing.filter((one) => one.level === level);
  const named = new Map(
    wanted.map((one) => [`${one.level ?? level} ${one.header}`, one]),
  );
  const out = [];

  for (const one of wanted) {
    const at = one.level ?? level;
    if (
      !one.required ||
      standing.some((held) => held.header === one.header && held.level === at)
    )
      continue;
    out.push(fault(one.header, where, 1, `A ${kind} carries a ${one.header} chapter.`));
  }

  for (const held of standing) {
    if (named.has(`${held.level} ${held.header}`)) continue;
    if (spec.extraSections === false) {
      out.push(
        fault(
          held.header,
          where,
          held.line,
          `The ${kind} schema names no ${held.header} chapter.`,
        ),
      );
    }
  }

  const once = wanted.filter((one) => (one.level ?? level) === level);
  if (spec.order === "strict") out.push(...orderFaults(top, once, kind, where));
  out.push(...lastFaults(top, once, where));

  for (const held of standing) {
    const rule = named.get(`${held.level} ${held.header}`);
    if (rule) out.push(...sectionFaults(held, rule, note, where));
  }
  return out;
}

// [[spec/design_output/schema#three-keywords-name-a-step]]
export function chaptersWanted(sections, front, level) {
  const out = [];
  for (const one of sections ?? []) {
    const list = one["x-one-per"];
    if (!list) {
      out.push({ ...one, level: one.level ?? level });
      continue;
    }
    out.push(...chaptersOf(front?.[list], level, one));
  }
  return out;
}

function chaptersOf(list, level, rule, listed = false) {
  const out = [];
  for (const one of [list ?? []].flat()) {
    if (!one || typeof one !== "object") continue;
    const header = String(one.name ?? "").trim();
    if (!header) continue;
    out.push({
      ...rule,
      header,
      level,
      required: true,
      description: askedOf(one) || sayOf(one),
      form: String(one.form ?? ""),
      "x-fills": Boolean(one.form),
    });
    const asks =
      listed || [one.checklist ?? []].flat().some((it) => String(it ?? "").trim());
    let leaf = true;
    for (const value of Object.values(one)) {
      if (!Array.isArray(value)) continue;
      if (!value.some((it) => it && typeof it === "object" && it.name)) continue;
      out.push(...chaptersOf(value, level + 1, rule, asks));
      if (value === one.steps) leaf = false;
    }
    // [[spec/design_output/pull#the-fields-hold-their-forms]]
    if (asks && leaf && one.evidence) {
      out.push({
        ...rule,
        header: CHECKED,
        level: level + 1,
        required: false,
        description:
          "one line per item of the checklist, on how you take it into account",
        form: "checklist",
        "x-fills": true,
      });
    }
  }
  return out;
}

export const CHECKED = "checked";

// A step the engine parks carries its question in `asks`, and the chapter under it reads that question, so the ticket alone says what the step waits on. [[spec/design_output/pull#a-person-step-goes-in]]
function askedOf(one) {
  return String(one?.asks ?? "")
    .replace(/\s+/g, " ")
    .replace(/--+>/g, "->")
    .trim();
}

// [[spec/design_output/schema#the-render-follows-the-tree]]
function sayOf(one) {
  for (const key of ["does", "says"]) {
    if (String(one?.[key] ?? "").trim()) return String(one[key]).trim();
  }
  return "";
}

function orderFaults(standing, wanted, kind, where) {
  const order = wanted.map((one) => one.header);
  const held = standing.filter((one) => order.includes(one.header));
  const out = [];
  for (let i = 1; i < held.length; i++) {
    if (order.indexOf(held[i].header) > order.indexOf(held[i - 1].header)) continue;
    out.push(
      fault(
        held[i].header,
        where,
        held[i].line,
        `${held[i].header} stands after ${held[i - 1].header}, and a ${kind} note puts it first.`,
      ),
    );
  }
  return out;
}

function lastFaults(standing, wanted, where) {
  const out = [];
  for (const one of wanted) {
    if (one.position !== "last") continue;
    const at = standing.findIndex((held) => held.header === one.header);
    if (at < 0 || at === standing.length - 1) continue;
    out.push(
      fault(
        one.header,
        where,
        standing[at].line,
        `${one.header} closes this note, and ${standing[at + 1].header} stands after it.`,
      ),
    );
  }
  return out;
}

function sectionFaults(held, rule, note, where) {
  const out = [];
  const items = itemsIn(held.own);

  if (rule.list && !items.length) {
    out.push(
      fault(held.header, where, held.line, `${held.header} holds a list of items.`),
    );
  }
  if (rule.ordered && items.some((one) => !NUMBERED.test(one.said))) {
    out.push(
      fault(held.header, where, held.line, `${held.header} numbers every item.`),
    );
  }
  if (rule.maxItems && items.length > rule.maxItems) {
    out.push(
      fault(
        held.header,
        where,
        held.line + items[rule.maxItems].line,
        `A note holds ${rule.maxItems} items.`,
      ),
    );
  }
  if (rule.subsections) {
    out.push(...underFaults(held, rule, note, where));
  }
  if (rule.table) {
    out.push(...tableFaults(held, rule.table, note, where, itemsIn, fault));
  }
  return out;
}

function underFaults(held, rule, note, where) {
  const spec = rule.subsections ?? {};
  const level = spec.headingLevel ?? held.level + 1;
  const at = note.sections.indexOf(held);
  const out = [];
  const numbers = [];

  for (const one of note.sections.slice(at + 1)) {
    if (one.level <= held.level) break;
    if (one.level !== level) continue;
    const found = NUMBERED.exec(one.header);
    if (spec.numbered && !found) {
      out.push(
        fault(
          held.header,
          where,
          one.line,
          `A chapter under ${held.header} opens with the number of the item it argues.`,
        ),
      );
      continue;
    }
    if (found)
      numbers.push({
        said: Number(found[1]),
        line: one.line,
        header: one.header,
      });
  }

  if (spec.order !== "strict") return out;
  for (let i = 1; i < numbers.length; i++) {
    if (numbers[i].said > numbers[i - 1].said) continue;
    out.push(
      fault(
        held.header,
        where,
        numbers[i].line,
        `${numbers[i].header} stands after ${numbers[i - 1].header}, and the numbers run up.`,
      ),
    );
  }
  return out;
}

// [[spec/design_output/schema#a-comment-counts-toward-nothing]]
export function itemsIn(rows) {
  const out = [];
  let fenced = false;
  for (let i = 0; i < rows.length; i++) {
    if (FENCE.test(rows[i])) {
      fenced = !fenced;
      continue;
    }
    if (fenced) continue;
    const said = String(rows[i]).replace(COMMENT, "").trim();
    if (ITEM.test(said)) out.push({ said, line: i + 1 });
  }
  return out;
}

// [[spec/design_output/schema#a-placeholder-stands-at-warning]]
export function placeholderFaults(text, schema, where) {
  const rows = String(text ?? "").split(/\r?\n/);
  const note = readNote(text);
  const props = schema?.frontmatter?.properties ?? {};
  const out = [];

  for (const [key, line] of Object.entries(note.front.lines ?? {})) {
    const rule = props[key];
    if (!rule || rule.const !== undefined || Array.isArray(rule.enum)) continue;
    if (String(rows[line - 1] ?? "").trim() !== `${key}: ${minted(rule)}`) continue;
    out.push(left(where, line, key));
  }

  const named = new Map(
    chaptersWanted(
      schema?.body?.sections ?? [],
      note.front.said ?? {},
      schema?.body?.headingLevel ?? 1,
    )
      .filter((one) => one.description && one["x-fills"] !== false)
      .map((one) => [one.header, `<!-- ${one.description} -->`]),
  );
  for (const held of note.sections) {
    const said = named.get(held.header);
    if (!said) continue;
    const at = held.own.findIndex((line) => String(line).trim() === said);
    if (at < 0) continue;
    out.push(left(where, held.line + at + 1, held.header));
  }
  return out;
}

export function left(file, line, what) {
  return {
    file,
    rule: "Schema.Placeholder",
    line,
    column: 1,
    message: `${what} still carries the placeholder mint writes. Say what stands there.`,
    severity: LEFT,
  };
}

export function minted(rule) {
  if (rule?.const !== undefined) {
    return rule["x-link"] ? `[[${rule.const}]]` : String(rule.const);
  }
  if (Array.isArray(rule?.enum)) return String(rule.enum[0]);

  const said = String(rule?.description ?? "what goes here");
  if ([rule?.type].flat().includes("array")) return `["${said}"]`;
  return rule?.["x-link"] ? `[[${said}]]` : said;
}
